Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +504,483 @@ void AchievementManager::FilereaderClose(void* file_handle)
-  rc_api_destroy_login_response(&login_data);
-  return r_type;
-}
 
-AchievementManager::ResponseType AchievementManager::ResolveHash(const Hash& game_hash,
-                                                                 u32* game_id)
-{
-  rc_api_resolve_hash_response_t hash_data{};
-  std::string username, api_token;
+  std::string config_username = Config::Get(Config::RA_USERNAME);
+  if (config_username != user->username)
   {
-    std::lock_guard lg{m_lock};
-    username = Config::Get(Config::RA_USERNAME);
-    api_token = Config::Get(Config::RA_API_TOKEN);
+    if (Common::CaseInsensitiveEquals(config_username, user->username))
+    {
+      INFO_LOG_FMT(ACHIEVEMENTS,
+                   "Case mismatch between site {} and local {}; updating local config.",
+                   user->username, Config::Get(Config::RA_USERNAME));
+      Config::SetBaseOrCurrent(Config::RA_USERNAME, user->username);
+    }
+    else
+    {
+      INFO_LOG_FMT(ACHIEVEMENTS, "Attempted to login prior user {}; current user is {}.",
+                   user->username, Config::Get(Config::RA_USERNAME));
+      rc_client_logout(client);
+      return;
+    }
   }
-  rc_api_resolve_hash_request_t resolve_hash_request = {
-      .username = username.c_str(), .api_token = api_token.c_str(), .game_hash = game_hash.data()};
-  ResponseType r_type = Request<rc_api_resolve_hash_request_t, rc_api_resolve_hash_response_t>(
-      resolve_hash_request, &hash_data, rc_api_init_resolve_hash_request,
-      rc_api_process_resolve_hash_response);
-  if (r_type == ResponseType::SUCCESS)
+  INFO_LOG_FMT(ACHIEVEMENTS, "Successfully logged in {} to RetroAchievements server.",
+               user->username);
+  std::lock_guard lg{AchievementManager::GetInstance().GetLock()};
+  Config::SetBaseOrCurrent(Config::RA_API_TOKEN, user->token);
+  AchievementManager::GetInstance().FetchPlayerBadge();
+}
+
+void AchievementManager::FetchBoardInfo(AchievementId leaderboard_id)
+{
+  u32* callback_data_1 = new u32(leaderboard_id);
+  u32* callback_data_2 = new u32(leaderboard_id);
+  rc_client_begin_fetch_leaderboard_entries(m_client, leaderboard_id, 1, 4,
+                                            LeaderboardEntriesCallback, callback_data_1);
+  rc_client_begin_fetch_leaderboard_entries_around_user(
+      m_client, leaderboard_id, 4, LeaderboardEntriesCallback, callback_data_2);
+}
+
+void AchievementManager::LeaderboardEntriesCallback(int result, const char* error_message,
+                                                    rc_client_leaderboard_entry_list_t* list,
+                                                    rc_client_t* client, void* userdata)
+{
+  if (result != RC_OK)
   {
-    *game_id = hash_data.game_id;
-    INFO_LOG_FMT(ACHIEVEMENTS, "Hashed game ID {} for RetroAchievements.", *game_id);
+    WARN_LOG_FMT(ACHIEVEMENTS, "Failed to fetch leaderboard entries.");
+    return;
   }
-  else
+
+  u32 leaderboard_id = *reinterpret_cast<u32*>(userdata);
+  delete userdata;
```

Comment:
Getting a new warning from this line: "cannot delete expression with pointer-to-'void' type 'void *' [-Wdelete-incomplete]". You're probably getting memory leaks from this @LillyJadeKatrin.